mantidproject/mantid
# Load grouping from IDF if not present in file

`LoadMuonNexus` should load grouping from IDF if there is no entry in the NeXus file.

The property _DetectorGroupingTable_ in `LoadMuonNexus-v1` is an output property that returns
a grouping table. If there is no grouping information in the NeXus file but this property is specified,
first try loading grouping from the IDF. If that fails, use a dummy grouping. (In either of these cases,
a log message specifies what has been done).

Also deal with files in which there is a grouping entry, but it is all zeros or otherwise invalid.

Use case: this enables such files to be used in ALC (and elsewhere) in the same way as MuonAnalysis (which already handles bad grouping in this way).

Added a note to the algorithm documentation, and two unit tests for missing and zero grouping.

**To test:**

Files: can be found at `\\olympic\babylon5\Scratch\TomPerkins\Data\grouping`.
`EMU00019489.nxs` has a grouping entry, but it's all zero.
`CHRONUS0003422.nxs` has no grouping entry at all.
(You can see this using HDFView, under `run/histogram_data_1/grouping`).
- Run `LoadMuonNexus` on these files, specifying a name for _DetectorGroupingTable_.
  - A TableWorkspace should be produced with the supplied name. It should have the detector grouping in it.
- Check that the new unit tests cover these two cases.

Fixes #14085.

[Release notes](https://github.com/mantidproject/mantid/blob/14085_missing_grouping/docs/source/release/v3.7.0/muon.rst#algorithms) have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
